domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,354 @@
+	declareEqualsClause("root_storage_type", filter.RootStorageType)
+
+	var inputArgs []any
+	if hasFilter {
+		inputArgs = append(inputArgs, filter)
+	}
+
+	// Handle  IN clauses, which is are bit different than equals clauses.
+	if len(filter.Versions) > 0 {
+		clauses = append(clauses, `cloud_image_metadata.version IN ($Versions[:])`)
+		inputArgs = append(inputArgs, filter.Versions)
+	}
+	if len(filter.Arches) > 0 {
+		clauses = append(clauses, `arch.architecture_name IN ($Arches[:])`)
+		inputArgs = append(inputArgs, filter.Arches)
+	}
+
+	findMetadataQuery := fmt.Sprintf(`
+WITH arch(id, architecture_name) AS (
+	SELECT id, name FROM architecture
+)
+SELECT &outputMetadata.* 
+FROM cloud_image_metadata
+JOIN arch ON cloud_image_metadata.architecture_id = arch.id
+WHERE %s;`, strings.Join(clauses, ` AND `))
+
+	findMetadataStmt, err := sqlair.Prepare(findMetadataQuery, append([]any{outputMetadata{}}, inputArgs...)...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var metadatas []outputMetadata
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, findMetadataStmt, inputArgs...).GetAll(&metadatas)
+		if dberrors.IsErrNotFound(err) {
+			return cloudmetadataerrors.NotFound
+		}
+		return err
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	result := make([]cloudimagemetadata.Metadata, len(metadatas))
+	for i, m := range metadatas {
+		result[i] = cloudimagemetadata.Metadata{
+			MetadataAttributes: cloudimagemetadata.MetadataAttributes{
+				Stream:          m.Stream,
+				Region:          m.Region,
+				Version:         m.Version,
+				Arch:            m.ArchitectureName,
+				VirtType:        m.VirtType,
+				RootStorageType: m.RootStorageType,
+				RootStorageSize: m.RootStorageSize,
+				Source:          m.Source,
+			},
+			Priority:     m.Priority,
+			ImageId:      m.ImageID,
+			CreationTime: m.CreatedAt,
+		}
+	}
+	return result, nil
+}
+
+// AllCloudImageMetadata retrieves all cloud image metadata from the database without applying any filter criteria.
+// Returns a slice of [cloudimagemetadata.Metadata] and any error encountered during the retrieval process.
+func (s *State) AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error) {
+	result, err := s.FindMetadata(ctx, cloudimagemetadata.MetadataFilter{})
+	if errors.Is(err, cloudmetadataerrors.NotFound) {
+		return nil, nil
+	}
+	return result, errors.Trace(err)
+}
+
+// tryCleanUpExpiredMetadata removes metadata records from the cloud_image_metadata table that are older than [ExpirationDelay].
+//
+// Custom images doesn't expire.
+func (s *State) tryCleanUpExpiredMetadata(ctx context.Context) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteQuery := fmt.Sprintf(`
+DELETE FROM cloud_image_metadata 
+WHERE source != '%s' AND  
+created_at < '%s';`, CustomSource, s.clock.Now().Add(-ExpirationDelay).Format(time.DateTime))
```

Comment:
Use a type here, like you've done in the filter.

Reply:
There is no possible SQL injectio here, but I agree that in the future, expiration time can somewhat been a user parameter and changing it may avoid futur security issue. Will do it.

Reply:
Ahah i fixed it but in FindMetadata. Sorry fix that asap